### DISCUSSION TITLE
How to Apply a Nodal Transformation to Densify Mesh in Gridap?

### DISCUSSION THREAD

**Discussion by JCHostos** [gridap/Gridap.jl#1035] [General]
Hi everyone,

I am working with Gridap.jl and generated a simple structured mesh using the following code:

domain = (0,1,0,1)
partition = (20,20)
model = CartesianDiscreteModel(domain, partition)

This creates a structured 20x20 quadrilateral mesh in the domain [0,1]×[0,1]. Now, I would like to modify the structure "model" to densify the mesh towards the edges using the following transformation:

(xn,yn)= 0.5 + 0.5 * tanh(eps * (2 * (x,y) - 1)) / tanh(eps)

where (xn,yn) are the new nodal coordinates after applying the transformation to the original coordinates (x,y), epsilon is the densification parameter.

​My Question:

How can I modify the model structure in Gridap to change only the nodal coordinates according to this transformation, while keeping the connectivity and other aspects of the mesh unchanged?. I adding pictures of what I want for clarity:

I want to past from this:

![Uniform](https://github.com/user-attachments/assets/c64cf9b7-878b-4da1-8f85-988b676f952a)

to this:

![Densified](https://github.com/user-attachments/assets/fbf5fce5-cb0e-479c-b40d-8b88632d9627)

**Comment by JordiManyer** [comment]
```julia
f(X) = .... where X = (x,y)
model = CartesianDiscreteModel(domain,partition;map=f)
```

**Reply by JCHostos** [reply]
Thank you for your answer, Jordi.

I try something like this:
function Clustering(X)
  alpha=2.0
  x, y = X
    xn = 0.5 + 0.5 * tanh(alpha * (2 * x - 1)) / tanh(alpha)
    yn = 0.5 + 0.5 * tanh(alpha * (2 * y - 1)) / tanh(alpha)
  return (xn, yn)
end
model = CartesianDiscreteModel(domain,partition;map=Clustering)

but it did not work. It is maybe too basic, but I am not a programmer so I will need a little more guidance. I am very new working on this powerfull Gridap enviroment.

Beforehand, thank you very much

**Reply by JordiManyer** [reply]
Our `Point` type, used for coordinates, is not iterable. I.e you cannot do this `x, y = X`.
Instead, do this: 
```
x, y = X[1], X[2]
```
Also, you need to return the same type, i.e instead of `(xn,yn)`, return `Point(xn,yn)`. 

For future reference, saying something "doesn't work" is quite useless to anyone trying to help. Posting the terminal output of the error usually makes this kind of interaction go quicker. :)

**Reply by JCHostos** [reply]
Thank you very much, Jordi.

I will show the terminal output of the error for futuer opportunities. But now, thanks to you, it worked. So, this is that I have to show:

![Captura de pantalla 2024-10-10 012533](https://github.com/user-attachments/assets/0d0bf72d-cb88-42af-be97-025c6f32b081)

Muchas gracias!!!
